mm2src/coins/tendermint/tendermint_tx_history_v2.rs

```diff
@@ -0,0 +1,754 @@
+                timestamp: common::now_ms() / 1000,
+                kmd_rewards: None,
+                transaction_type: Default::default(),
+            }))
+        }
+
+        async fn parse_create_htlc_tx_details<Coin, Storage>(
+            args: ParseTxDetailsArgs<Coin, Storage, '_>,
+        ) -> Result<Option<TransactionDetails>, StopReason>
+        where
+            Coin: TendermintTxHistoryOps,
+            Storage: TxHistoryStorage,
+        {
+            let htlc_tx =
+                CreateHtlcProtoRep::decode(args.tx_body_msg).map_err(|e| StopReason::Marshaling(e.to_string()))?;
+
+            let coin_amount = htlc_tx
+                .amount
+                .first()
+                .ok_or_else(|| StopReason::Marshaling("amount can't be empty".to_string()))?;
+
+            let denom = coin_amount.denom.to_lowercase();
+            let tx_coin_ticker = match args.active_assets.get(&denom) {
+                Some(t) => t,
+                None => return Ok(None),
+            };
+
+            let wallet_id = WalletId::new(tx_coin_ticker.replace('-', "_"));
+            if matches!(
+                args.storage.get_tx_from_history(&wallet_id, &args.internal_id).await,
+                Ok(Some(_))
+            ) {
+                log::debug!("Tx '{}' already exists in tx_history. Skipping it.", &args.tx_hash);
+                return Ok(None);
+            } else {
+                log::debug!("Adding tx: '{}' to tx_history.", &args.tx_hash);
+            }
+
+            let address = args.coin.my_address().expect("my_address can't fail");
+            let tx_amounts = get_tx_amounts(coin_amount, args.coin.decimals(), htlc_tx.sender == address)
+                .map_err(StopReason::Marshaling)?;
+
+            Ok(Some(TransactionDetails {
+                from: vec![htlc_tx.sender],
+                to: vec![htlc_tx.to.clone()],
+                total_amount: tx_amounts.total,
+                spent_by_me: tx_amounts.spent_by_me,
+                received_by_me: tx_amounts.received_by_me,
+                my_balance_change: tx_amounts.my_balance_change,
+                tx_hash: args.tx_hash,
+                tx_hex: args.tx_body_msg.into(),
+                fee_details: Some(args.fee_details),
+                block_height: args.block_height,
+                coin: tx_coin_ticker.to_string(),
+                internal_id: args.internal_id,
+                timestamp: common::now_ms() / 1000,
+                kmd_rewards: None,
+                transaction_type: Default::default(),
+            }))
+        }
+
+        async fn fetch_and_insert_txs<Coin, Storage>(
+            coin: &Coin,
+            storage: &Storage,
+            query: String,
+            active_assets: &HashMap<String, String>,
+        ) -> Result<(), Stopped<Coin, Storage>>
+        where
+            Coin: TendermintTxHistoryOps,
+            Storage: TxHistoryStorage,
+        {
+            let mut page = 1;
```

Comment:
Do I understand correctly that each time when the balance is changed, we request the whole TX history and every transaction details? I think it's suboptimal, but it can be improved at the next iteration

Reply:
Yes, I forgot to add TODO note about this. We should save the block height and continue from there on each trigger. I will implement it in current PR if I will have enough time.

Reply:
Implemented it. Now we don't re-fetch transactions from beginning.